src/kde/quick_index.cc

```diff
@@ -0,0 +1,813 @@
+                                index[2] * coarse_bin_resolution * coarse_bin_resolution;
+    coarse_index_map[global_index].push_back(locIndex);
+    locIndex++;
+  }
+
+  // Now we need to build up ghost location map data for domain decomposed mode
+  if (domain_decomposed) {
+    // temporary cast of the nodes to prevent conversion warnings
+    const size_t nodes = static_cast<size_t>(rtt_c4::nodes());
+    const size_t node = static_cast<size_t>(rtt_c4::node());
+
+    // build list of local bins based on the local bounds
+    local_bins = window_coarse_index_list(local_bounding_box_min, local_bounding_box_max);
+
+    // build a global map for number of entries into the global bins on each processor
+    // creates a (nbins**dim)*nranks sized array
+    // NOTE: If this gets to big we could stride over a subset of coarse bins
+    // and do multiple iterations of mpi communication to build up the map
+    size_t nbins = coarse_bin_resolution;
+    for (size_t d = 1; d < dim; d++)
+      nbins *= coarse_bin_resolution;
+
+    std::vector<int> global_index_per_bin_per_proc(nbins * nodes, 0UL);
+    for (auto mapItr = coarse_index_map.begin(); mapItr != coarse_index_map.end(); mapItr++) {
+      size_t gipbpp_index = mapItr->first + nbins * node;
+      // must cast to an int to accomidate mpi int types.
+      global_index_per_bin_per_proc[gipbpp_index] = static_cast<int>(mapItr->second.size());
+    }
+    rtt_c4::global_sum(&global_index_per_bin_per_proc[0], nbins * nodes);
+
+    // calculate local ghost buffer size
+    local_ghost_buffer_size = 0;
+    for (size_t proc = 0; proc < nodes; proc++) {
+      for (auto binItr = local_bins.begin(); binItr != local_bins.end(); binItr++) {
+        if (node != proc) {
+          size_t gipbpp_index = *binItr + nbins * proc;
+          // build up the local ghost index map
+          for (int i = 0; i < global_index_per_bin_per_proc[gipbpp_index]; i++)
+            local_ghost_index_map[*binItr].push_back(local_ghost_buffer_size + i);
+          // accumulate the total ghost points
+          local_ghost_buffer_size += global_index_per_bin_per_proc[gipbpp_index];
+        }
+      }
+    }
+
+    std::vector<int> global_need_bins_per_proc(nbins * nodes, 0UL);
+    // global need bins
+    for (auto binItr = local_bins.begin(); binItr != local_bins.end(); binItr++) {
+      global_need_bins_per_proc[*binItr + nbins * node] += 1;
+    }
+    rtt_c4::global_sum(&global_need_bins_per_proc[0], nbins * nodes);
+
+    // Build a global list of buffer sizes
+    std::vector<int> proc_ghost_buffer_size(nodes, 0);
+    proc_ghost_buffer_size[node] = local_ghost_buffer_size;
+    rtt_c4::global_sum(&proc_ghost_buffer_size[0], nodes);
+
+    // calculate the put map so each node knows which processor to send data
+    // and where to index that data
+    for (int rec_proc = 0; rec_proc < rtt_c4::nodes(); rec_proc++) {
+      // calculating the offset SUCKS!!! If anyone can find a better way please help.
```

Comment:
I'll try to look closer, but offsets like this can be often be calculated with the `partial_sum` command (https://en.cppreference.com/w/cpp/algorithm/partial_sum). If you need to do it in parallel, consider the MPI scan operation (https://www.open-mpi.org/doc/v4.1/man3/MPI_Scan.3.php)

Reply:
I'm going to balk on this change for now. I'm not 100% sure how to do this. For now I've added a performance note to look at this. Is that alright with you @alexrlongne.